// codechain-sdk/lib/rpc/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var codechain_primitives_1 = require("codechain-primitives");
var chain_1 = require("./chain");
var AccountRpc = /** @class */ (function () {
    /**
     * @hidden
     */
    function AccountRpc(rpc) {
        this.rpc = rpc;
    }
    /**
     * Gets a list of accounts.
     * @returns A list of accounts
     */
    AccountRpc.prototype.getList = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.rpc
                .sendRpcRequest("account_getList", [])
                .then(function (accounts) {
                try {
                    if (Array.isArray(accounts)) {
                        resolve(accounts.map(function (account) {
                            return codechain_primitives_1.PlatformAddress.ensure(account).toString();
                        }));
                    }
                    else {
                        reject(Error("Expected account_getList to return an array but it returned " + accounts));
                    }
                }
                catch (e) {
                    reject(Error("Expected account_getList to return an array of PlatformAddress string, but an error occurred: " + e.toString()));
                }
            })
                .catch(reject);
        });
    };
    /**
     * Creates a new account.
     * @param passphrase A passphrase to be used by the account owner
     * @returns An account
     */
    AccountRpc.prototype.create = function (passphrase) {
        var _this = this;
        if (passphrase && typeof passphrase !== "string") {
            throw Error("Expected the first argument to be a string but given " + passphrase);
        }
        return new Promise(function (resolve, reject) {
            _this.rpc
                .sendRpcRequest("account_create", [passphrase])
                .then(function (account) {
                try {
                    resolve(codechain_primitives_1.PlatformAddress.ensure(account).toString());
                }
                catch (e) {
                    reject(Error("Expected account_create to return PlatformAddress string but an error occurred: " + e.toString()));
                }
            })
                .catch(reject);
        });
    };
    /**
     * Imports a secret key and add the corresponding account.
     * @param secret H256 or hexstring for 256-bit private key
     * @param passphrase A passphrase to be used by the account owner
     * @returns The account
     */
    AccountRpc.prototype.importRaw = function (secret, passphrase) {
        var _this = this;
        if (!codechain_primitives_1.H256.check(secret)) {
            throw Error("Expected the first argument to be an H256 value but found " + secret);
        }
        if (passphrase && typeof passphrase !== "string") {
            throw Error("Expected the second argument to be a string but found " + passphrase);
        }
        return new Promise(function (resolve, reject) {
            _this.rpc
                .sendRpcRequest("account_importRaw", [
                "0x" + codechain_primitives_1.H256.ensure(secret).value,
                passphrase
            ])
                .then(function (account) {
                try {
                    resolve(codechain_primitives_1.PlatformAddress.ensure(account).toString());
                }
                catch (e) {
                    reject(Error("Expected account_importRaw to return PlatformAddress string but an error occurred: " + e.toString()));
                }
            })
                .catch(reject);
        });
    };
    /**
     * Calculates the account's signature for a given message.
     * @param messageDigest A message to sign
     * @param address A platform address
     * @param passphrase The account's passphrase
     */
    AccountRpc.prototype.sign = function (messageDigest, address, passphrase) {
        var _this = this;
        if (!codechain_primitives_1.H256.check(messageDigest)) {
            throw Error("Expected the first argument to be an H256 value but found " + messageDigest);
        }
        if (!codechain_primitives_1.PlatformAddress.check(address)) {
            throw Error("Expected the second argument to be a PlatformAddress value but found " + address);
        }
        if (passphrase && typeof passphrase !== "string") {
            throw Error("Expected the third argument to be a string but found " + passphrase);
        }
        return new Promise(function (resolve, reject) {
            _this.rpc
                .sendRpcRequest("account_sign", [
                "0x" + codechain_primitives_1.H256.ensure(messageDigest).value,
                codechain_primitives_1.PlatformAddress.ensure(address).toString(),
                passphrase
            ])
                .then(function (result) {
                if (typeof result === "string" &&
                    result.match(/0x[0-9a-f]{130}/)) {
                    return resolve(result);
                }
                reject(Error("Expected account_sign to return a 65 byte hexstring but it returned " + result));
            })
                .catch(reject);
        });
    };
    /**
     * Sends a transaction with the account's signature.
     * @param params.tx A tx to send
     * @param params.account The platform account to sign the tx
     * @param params.passphrase The account's passphrase
     */
    AccountRpc.prototype.sendTransaction = function (params) {
        var tx = params.tx, fee = params.fee, account = params.account, passphrase = params.passphrase;
        if (!codechain_primitives_1.PlatformAddress.check(account)) {
            throw Error("Expected the account param to be a PlatformAddress value but found " + account);
        }
        if (passphrase && typeof passphrase !== "string") {
            throw Error("Expected the passphrase param to be a string but found " + passphrase);
        }
        if (fee) {
            tx.setFee(fee);
        }
        if (tx.fee() == null) {
            tx.setFee(chain_1.getMinimumFee(tx));
        }
        return this.rpc
            .sendRpcRequest("account_sendTransaction", [
            tx.toJSON(),
            codechain_primitives_1.PlatformAddress.ensure(account).toString(),
            passphrase
        ])
            .then(function (result) {
            return {
                hash: codechain_primitives_1.H256.ensure(result.hash),
                seq: result.seq
            };
        });
    };
    /**
     * Unlocks the account.
     * @param address A platform address
     * @param passphrase The account's passphrase
     * @param duration Time to keep the account unlocked. The default value is 300(seconds). Passing 0 unlocks the account indefinitely.
     */
    AccountRpc.prototype.unlock = function (address, passphrase, duration) {
        var _this = this;
        if (!codechain_primitives_1.PlatformAddress.check(address)) {
            throw Error("Expected the first argument to be a PlatformAddress value but found " + address);
        }
        if (passphrase && typeof passphrase !== "string") {
            throw Error("Expected the second argument to be a string but found " + passphrase);
        }
        if (duration !== undefined &&
            (typeof duration !== "number" ||
                !Number.isInteger(duration) ||
                duration < 0)) {
            throw Error("Expected the third argument to be non-negative integer but found " + duration);
        }
        return new Promise(function (resolve, reject) {
            _this.rpc
                .sendRpcRequest("account_unlock", [
                codechain_primitives_1.PlatformAddress.ensure(address).toString(),
                passphrase || "",
                duration
            ])
                .then(function (result) {
                if (result === null) {
                    return resolve(null);
                }
                reject(Error("Expected account_unlock to return null but it returned " + result));
            })
                .catch(reject);
        });
    };
    /**
     * Changes the passpharse of the account
     * @param address A platform address
     * @param oldPassphrase The account's current passphrase
     * @param newPassphrase The new passphrase for the account
     */
    AccountRpc.prototype.changePassword = function (address, oldPassphrase, newPassphrase) {
        var _this = this;
        if (oldPassphrase && typeof oldPassphrase !== "string") {
            throw Error("Expected the second argument to be a string but given " + oldPassphrase);
        }
        if (newPassphrase && typeof newPassphrase !== "string") {
            throw Error("Expected the second argument to be a string but given " + newPassphrase);
        }
        return new Promise(function (resolve, reject) {
            _this.rpc
                .sendRpcRequest("account_changePassword", [
                address,
                oldPassphrase,
                newPassphrase
            ])
                .then(function (result) {
                if (result == null) {
                    return resolve(null);
                }
                reject(Error("Expected account_changePassword to return null but it returned " + result));
            })
                .catch(reject);
        });
    };
    return AccountRpc;
}());
exports.AccountRpc = AccountRpc;
