// codechain-keystore/lib/logic/error.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ErrorCode;
(function (ErrorCode) {
    ErrorCode[ErrorCode["Unknown"] = 0] = "Unknown";
    ErrorCode[ErrorCode["NoSuchKey"] = 1] = "NoSuchKey";
    ErrorCode[ErrorCode["NoSuchSeedHash"] = 2] = "NoSuchSeedHash";
    ErrorCode[ErrorCode["DecryptionFailed"] = 3] = "DecryptionFailed";
    ErrorCode[ErrorCode["DBError"] = 4] = "DBError";
    ErrorCode[ErrorCode["WrongSeedLength"] = 5] = "WrongSeedLength";
    ErrorCode[ErrorCode["WrongMnemonicString"] = 6] = "WrongMnemonicString";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
var KeystoreError = /** @class */ (function (_super) {
    __extends(KeystoreError, _super);
    function KeystoreError(code) {
        var _this = _super.call(this, ErrorCode[code]) || this;
        _this.code = code;
        _this.codeName = ErrorCode[code];
        _this.name = "KeystoreError";
        return _this;
    }
    return KeystoreError;
}(Error));
exports.KeystoreError = KeystoreError;
