// codechain-keystore/lib/model/hdkeys.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var codechain_primitives_1 = require("codechain-primitives");
var _ = require("lodash");
var error_1 = require("../logic/error");
var hdstorage_1 = require("../logic/hdstorage");
var keytypes_1 = require("./keytypes");
var Mnemonic = require("bitcore-mnemonic");
var Random = Mnemonic.bitcore.crypto.Random;
// Seed: Seed(entropy) of a mnemonic code. Not a direct seed of a HD wallet.
// See the definition:
// https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki
function getSeedHashes(context) {
    return __awaiter(this, void 0, void 0, function () {
        var rows;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, context.db
                        .get(keytypes_1.getTableName(keytypes_1.KeyType.HDWSeed))
                        .value()];
                case 1:
                    rows = _a.sent();
                    return [2 /*return*/, _.map(rows, function (storage) { return storage.seedHash; })];
            }
        });
    });
}
exports.getSeedHashes = getSeedHashes;
function exportSeed(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretSeedStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchSeedHash);
                    }
                    return [4 /*yield*/, hdstorage_1.decode(secret, params.passphrase)];
                case 2:
                    _a.sent();
                    return [2 /*return*/, secret];
            }
        });
    });
}
exports.exportSeed = exportSeed;
function exportRawSeed(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretSeedStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchSeedHash);
                    }
                    return [2 /*return*/, hdstorage_1.decode(secret, params.passphrase)];
            }
        });
    });
}
exports.exportRawSeed = exportRawSeed;
function exportMnemonic(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret, seed, code;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretSeedStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchSeedHash);
                    }
                    return [4 /*yield*/, hdstorage_1.decode(secret, params.passphrase)];
                case 2:
                    seed = _a.sent();
                    code = new Mnemonic(Buffer.from(seed, "hex"));
                    return [2 /*return*/, code.toString()];
            }
        });
    });
}
exports.exportMnemonic = exportMnemonic;
function importSeed(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var seed;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, hdstorage_1.decode(params.secret, params.passphrase)];
                case 1:
                    seed = _a.sent();
                    return [2 /*return*/, importSeedToDB(context, {
                            seed: seed,
                            passphrase: params.passphrase,
                            meta: params.secret.meta
                        })];
            }
        });
    });
}
exports.importSeed = importSeed;
function importRawSeed(context, params) {
    return importSeedToDB(context, params);
}
exports.importRawSeed = importRawSeed;
function importMnemonic(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var mnemonic, wordlist, words, bin, _i, words_1, word, ind, cs, hashBits, nonhashBits, buf, i, expectedHashBits, seed;
        return __generator(this, function (_a) {
            mnemonic = params.mnemonic;
            wordlist = Mnemonic._getDictionary(mnemonic);
            if (!wordlist) {
                throw new error_1.KeystoreError(error_1.ErrorCode.WrongMnemonicString);
            }
            words = mnemonic.split(" ");
            bin = "";
            for (_i = 0, words_1 = words; _i < words_1.length; _i++) {
                word = words_1[_i];
                ind = wordlist.indexOf(word);
                if (ind < 0) {
                    throw new error_1.KeystoreError(error_1.ErrorCode.WrongMnemonicString);
                }
                bin = bin + ("00000000000" + ind.toString(2)).slice(-11);
            }
            cs = bin.length / 33;
            hashBits = bin.slice(-cs);
            nonhashBits = bin.slice(0, bin.length - cs);
            buf = new Buffer(nonhashBits.length / 8);
            for (i = 0; i < nonhashBits.length / 8; i++) {
                buf.writeUInt8(parseInt(bin.slice(i * 8, (i + 1) * 8), 2), i);
            }
            expectedHashBits = Mnemonic._entropyChecksum(buf);
            if (expectedHashBits !== hashBits) {
                throw new error_1.KeystoreError(error_1.ErrorCode.WrongMnemonicString);
            }
            seed = buf.toString("hex");
            return [2 /*return*/, importSeedToDB(context, __assign({}, params, { seed: seed }))];
        });
    });
}
exports.importMnemonic = importMnemonic;
function createSeed(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, seedLength, seed;
        return __generator(this, function (_b) {
            _a = params.seedLength, seedLength = _a === void 0 ? 128 : _a;
            if (seedLength % 32 !== 0 || seedLength < 128) {
                throw new error_1.KeystoreError(error_1.ErrorCode.WrongSeedLength);
            }
            seed = Random.getRandomBuffer(seedLength / 8).toString("hex");
            return [2 /*return*/, importSeedToDB(context, __assign({}, params, { seed: seed }))];
        });
    });
}
exports.createSeed = createSeed;
function deleteSeed(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretSeedStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        return [2 /*return*/, false];
                    }
                    return [4 /*yield*/, removeKey(context, params)];
                case 2:
                    _a.sent();
                    return [2 /*return*/, true];
            }
        });
    });
}
exports.deleteSeed = deleteSeed;
function getPrivateKeyFromSeed(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret, seed, code, masterKey, derivedKey, privateKey;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretSeedStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchSeedHash);
                    }
                    return [4 /*yield*/, hdstorage_1.decode(secret, params.passphrase)];
                case 2:
                    seed = _a.sent();
                    code = new Mnemonic(Buffer.from(seed, "hex"));
                    masterKey = code.toHDPrivateKey();
                    derivedKey = masterKey.derive(params.path);
                    privateKey = derivedKey.privateKey.toString();
                    return [2 /*return*/, privateKey];
            }
        });
    });
}
exports.getPrivateKeyFromSeed = getPrivateKeyFromSeed;
function getPublicKeyFromSeed(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var privateKey, publicKey;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getPrivateKeyFromSeed(context, params)];
                case 1:
                    privateKey = _a.sent();
                    publicKey = codechain_primitives_1.getPublicFromPrivate(privateKey);
                    return [2 /*return*/, publicKey];
            }
        });
    });
}
exports.getPublicKeyFromSeed = getPublicKeyFromSeed;
function signFromSeed(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var privateKey, _a, r, s, v, sig;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, getPrivateKeyFromSeed(context, params)];
                case 1:
                    privateKey = _b.sent();
                    _a = codechain_primitives_1.signEcdsa(params.message, privateKey), r = _a.r, s = _a.s, v = _a.v;
                    sig = "" + _.padStart(r, 64, "0") + _.padStart(s, 64, "0") + _.padStart(v.toString(16), 2, "0");
                    return [2 /*return*/, sig];
            }
        });
    });
}
exports.signFromSeed = signFromSeed;
function importSeedToDB(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var passphrase, meta, secret, rows;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    passphrase = params.passphrase || "";
                    meta = params.meta || "{}";
                    return [4 /*yield*/, hdstorage_1.encode(params.seed, passphrase, meta)];
                case 1:
                    secret = _a.sent();
                    rows = context.db.get(keytypes_1.getTableName(keytypes_1.KeyType.HDWSeed));
                    return [4 /*yield*/, rows.push(secret).write()];
                case 2:
                    _a.sent();
                    return [2 /*return*/, secret.seedHash];
            }
        });
    });
}
function getSecretSeedStorage(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var collection, secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    collection = context.db.get(keytypes_1.getTableName(keytypes_1.KeyType.HDWSeed));
                    return [4 /*yield*/, collection
                            .find(function (secretStorage) {
                            return secretStorage.seedHash === params.seedHash;
                        })
                            .value()];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        return [2 /*return*/, null];
                    }
                    return [2 /*return*/, secret];
            }
        });
    });
}
function removeKey(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var collection;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    collection = context.db.get(keytypes_1.getTableName(keytypes_1.KeyType.HDWSeed));
                    return [4 /*yield*/, collection
                            .remove(function (secret) { return secret.seedHash === params.seedHash; })
                            .write()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function getMeta(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretSeedStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchKey);
                    }
                    return [2 /*return*/, secret.meta];
            }
        });
    });
}
exports.getMeta = getMeta;
function save(context) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, context.db.get(keytypes_1.getTableName(keytypes_1.KeyType.HDWSeed)).value()];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.save = save;
function load(context, value) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, context.db.set(keytypes_1.getTableName(keytypes_1.KeyType.HDWSeed), value).write()];
        });
    });
}
exports.load = load;
function clear(context) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, context.db.unset(keytypes_1.getTableName(keytypes_1.KeyType.HDWSeed)).write()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.clear = clear;
