// codechain-keystore/lib/model/keys.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var codechain_primitives_1 = require("codechain-primitives");
var _ = require("lodash");
var error_1 = require("../logic/error");
var storage_1 = require("../logic/storage");
var keytypes_1 = require("./keytypes");
function getKeys(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var rows;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, context.db
                        .get(keytypes_1.getTableName(params.keyType))
                        .value()];
                case 1:
                    rows = _a.sent();
                    return [2 /*return*/, _.map(rows, function (secret) { return secret.address; })];
            }
        });
    });
}
exports.getKeys = getKeys;
function getPublicKey(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret, privateKey;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        return [2 /*return*/, null];
                    }
                    return [4 /*yield*/, storage_1.decode(secret, params.passphrase)];
                case 2:
                    privateKey = _a.sent();
                    return [2 /*return*/, codechain_primitives_1.getPublicFromPrivate(privateKey)];
            }
        });
    });
}
exports.getPublicKey = getPublicKey;
function importRaw(context, params) {
    return createKeyFromPrivateKey(context, params);
}
exports.importRaw = importRaw;
function exportKey(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchKey);
                    }
                    return [4 /*yield*/, storage_1.decode(secret, params.passphrase)];
                case 2:
                    _a.sent(); // Throws an error if the passphrase is incorrect.
                    return [2 /*return*/, secret];
            }
        });
    });
}
exports.exportKey = exportKey;
function importKey(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var privateKey;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, storage_1.decode(params.secret, params.passphrase)];
                case 1:
                    privateKey = _a.sent();
                    return [2 /*return*/, importRaw(context, {
                            privateKey: privateKey,
                            passphrase: params.passphrase,
                            keyType: params.keyType,
                            meta: params.secret.meta
                        })];
            }
        });
    });
}
exports.importKey = importKey;
function createKey(context, params) {
    var privateKey = codechain_primitives_1.generatePrivateKey();
    return createKeyFromPrivateKey(context, __assign({}, params, { privateKey: privateKey }));
}
exports.createKey = createKey;
function createKeyFromPrivateKey(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var publicKey, passphrase, meta, secret, rows;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    publicKey = codechain_primitives_1.getPublicFromPrivate(params.privateKey);
                    passphrase = params.passphrase || "";
                    meta = params.meta || "{}";
                    return [4 /*yield*/, storage_1.encode(params.privateKey, params.keyType, passphrase, meta)];
                case 1:
                    secret = _a.sent();
                    rows = context.db.get(keytypes_1.getTableName(params.keyType));
                    return [4 /*yield*/, rows.push(secret).write()];
                case 2:
                    _a.sent();
                    return [2 /*return*/, keyFromPublicKey(params.keyType, publicKey)];
            }
        });
    });
}
function keyFromPublicKey(type, publicKey) {
    switch (type) {
        case keytypes_1.KeyType.Platform:
            return codechain_primitives_1.getAccountIdFromPublic(publicKey);
        case keytypes_1.KeyType.Asset:
            return codechain_primitives_1.H160.ensure(codechain_primitives_1.blake160(publicKey)).value;
        default:
            throw new Error("Invalid key type");
    }
}
exports.keyFromPublicKey = keyFromPublicKey;
function deleteKey(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        return [2 /*return*/, false];
                    }
                    return [4 /*yield*/, removeKey(context, params)];
                case 2:
                    _a.sent();
                    return [2 /*return*/, true];
            }
        });
    });
}
exports.deleteKey = deleteKey;
function getSecretStorage(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var collection, secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    collection = context.db.get(keytypes_1.getTableName(params.keyType));
                    return [4 /*yield*/, collection
                            .find(function (secretStorage) {
                            return secretStorage.address === params.key;
                        })
                            .value()];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        return [2 /*return*/, null];
                    }
                    return [2 /*return*/, secret];
            }
        });
    });
}
function removeKey(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var collection;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    collection = context.db.get(keytypes_1.getTableName(params.keyType));
                    return [4 /*yield*/, collection
                            .remove(function (secret) { return secret.address === params.key; })
                            .write()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function exportRawKey(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchKey);
                    }
                    return [2 /*return*/, storage_1.decode(secret, params.passphrase)];
            }
        });
    });
}
exports.exportRawKey = exportRawKey;
function sign(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret, privateKey, _a, r, s, v;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, getSecretStorage(context, params)];
                case 1:
                    secret = _b.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchKey);
                    }
                    return [4 /*yield*/, storage_1.decode(secret, params.passphrase)];
                case 2:
                    privateKey = _b.sent();
                    _a = codechain_primitives_1.signEcdsa(params.message, privateKey), r = _a.r, s = _a.s, v = _a.v;
                    return [2 /*return*/, "" + _.padStart(r, 64, "0") + _.padStart(s, 64, "0") + _.padStart(v.toString(16), 2, "0")];
            }
        });
    });
}
exports.sign = sign;
function getMeta(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        var secret;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, getSecretStorage(context, params)];
                case 1:
                    secret = _a.sent();
                    if (secret == null) {
                        throw new error_1.KeystoreError(error_1.ErrorCode.NoSuchKey);
                    }
                    return [2 /*return*/, secret.meta];
            }
        });
    });
}
exports.getMeta = getMeta;
function save(context, params) {
    return context.db.get(keytypes_1.getTableName(params.keyType)).value();
}
exports.save = save;
function load(context, value, params) {
    return context.db.set(keytypes_1.getTableName(params.keyType), value).write();
}
exports.load = load;
function clear(context, params) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, context.db.unset(keytypes_1.getTableName(params.keyType)).write()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.clear = clear;
